### PR TITLE
fixed some clippy warnings

### PR DESCRIPTION
**Pull request by llogiq**
Those are mostly style nits, but some things may even improve performance (like use of the `Entry` API or `Cow<'a, str>` instead of copying `String`s around).

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Wow, this is nice! Do you consider this PR ready or do you want to add some more patches?

Either way: Thank you very much!

**Comment by llogiq** [comment]
I'll try to add the rest later today if I manage to find the time.

**Comment by llogiq** [comment]
A few warnings remain, but those are fairly opinionated, so I'm OK with them for now. Cheers!

**Comment by matthiasbeyer** [comment]
Thank you very very much! Is it okay with you if I rewrite the commit messages (and add you in a `Suggested-by: ` commit tag)?

**Comment by llogiq** [comment]
Of course. Sorry, I should have asked (or looked around in the history)... I really want to be more disciplined around git history, but I only have so much time. :clock1130:

**Comment by matthiasbeyer** [comment]
@llogiq absolutely no problem! Non-Optimal commit messages are better than no commits/contributions, so thank you very much! I'll rewrite and merge this as soon as I can find some time!

**Comment by llogiq** [comment]
Btw. I'm still learning got, so how does one rewrite commit messages? Perhaps I can try and learn how to do it.

**Comment by matthiasbeyer** [comment]
@llogiq have a look at this: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History and feel free to ask me questions (here or by reaching out via email).

**Comment by llogiq** [comment]
OK, first question: Is it OK to rebase? I think I have read somewhere that one should never rebase already published branches?

**Comment by matthiasbeyer** [comment]
The point in rebasing is that you should never rewrite history where other people based work on. So generally: Don't rewrite history. For pull requests it is perfectly fine, and always if the maintainer of the project you request the pull to asks you to rewrite it.

**Comment by llogiq** [comment]
OK then I'll take a look at the commit history and try to reword following that example.

Question two: should I also squash?

**Comment by matthiasbeyer** [comment]
I have the policy "More detailed commits are better than one single monolithic one" - so please do not squash your commits!

**Comment by matthiasbeyer** [comment]
@llogiq Your PR has conflicts with master, can you rebase?

**Comment by llogiq** [comment]
I'm on mobile atm, probably this evening.

**Comment by matthiasbeyer** [comment]
No time pressure! :smile:

File: imag-link/src/main.rs [hunk 1]
```diff
@@ -81,16 +81,14 @@ fn handle_internal_linking(rt: &Runtime) {
...
-        for entry in cmd.value_of("list").unwrap().split(",") {
+        for entry in cmd.value_of("list").unwrap().split(',') {
             debug!("Listing for '{}'", entry);
             match get_entry_by_name(rt, entry) {
                 Ok(e) => {
                     e.get_internal_links()
                         .map(|links| {
-                            let mut i = 0;
-                            for link in links.iter().map(|l| l.to_str()).filter_map(|x| x) {
+                            for (i, link) in links.iter().map(|l| l.to_str()).filter_map(|x| x).enumerate() {
```
Comment (matthiasbeyer) [inline imag-link/src/main.rs]: Uh, I really like this change :+1:

File: imag-store/src/create.rs [hunk 2]
```diff
@@ -60,38 +60,32 @@ pub fn create(rt: &Runtime) {
...
+            |entry_matches| build_toml_header(entry_matches, EntryHeader::new()));
 
     create_with_content_and_header(rt, path, content, header)
 }
 
 fn create_from_source(rt: &Runtime, matches: &ArgMatches, path: &PathBuf) -> Result<()> {
     let content = matches
         .value_of("from-raw")
-        .ok_or(StoreError::new(StoreErrorKind::NoCommandlineCall, None))
+        .ok_or_else(|| StoreError::new(StoreErrorKind::NoCommandlineCall, None))
```
Comment (matthiasbeyer) [inline imag-store/src/create.rs]: What is the benefit of a closure here?

File: imag-store/src/create.rs [hunk 3]
```diff
@@ -60,38 +60,32 @@ pub fn create(rt: &Runtime) {
...
         .map(|raw_src| string_from_raw_src(raw_src));
```
Comment (matthiasbeyer) [inline imag-store/src/create.rs]: One could remove the variable here. I'd do this in a seperate PR if you don't want to.

File: libimagentrytag/src/tagable.rs [hunk 4]
```diff
@@ -7,21 +7,21 @@ use libimagstore::store::{Entry, EntryHeader, FileLockEntry};
...
+use tag::{Tag, TagSlice};
 use util::is_tag;
 
 use toml::Value;
 
 pub trait Tagable {
 
     fn get_tags(&self) -> Result<Vec<Tag>>;
-    fn set_tags(&mut self, ts: Vec<Tag>) -> Result<()>;
+    fn set_tags(&mut self, ts: &[Tag]) -> Result<()>;
```
Comment (matthiasbeyer) [inline libimagentrytag/src/tagable.rs]: What are the benefits of using a `&[]` here over a `Vec`?

File: libimagentrytag/src/util.rs [hunk 5]
```diff
@@ -1,5 +1,5 @@
 use regex::Regex;
 
-pub fn is_tag(s: &String) -> bool {
+pub fn is_tag(s: &str) -> bool {
     Regex::new("^[a-zA-Z]([a-zA-Z0-9_-]*)$").unwrap().captures(&s[..]).is_some()
```
Comment (matthiasbeyer) [inline libimagentrytag/src/util.rs]: I guess with the patch above we can pass plain `s` to the `captures()` function here, can't we?

**Comment by matthiasbeyer** [comment]
Besides my annotations, this looks really awesome! I also found some issues with the codebase I'm going to fix now (and merge after your PR, so you have less effort with rebasing and such).

File: libimagentrytag/src/tagable.rs (hunk 4 above)
Comment (llogiq) [inline libimagentrytag/src/tagable.rs]: `&[T]` is a strict superset of `&Vec<T>`, and allows to use arrays, arrayvecs, plain slices, etc.

File: imag-store/src/create.rs (hunk 2 above)
Comment (llogiq) [inline imag-store/src/create.rs]: Nothing. If you want, I'll remove it.

File: libimagentrytag/src/util.rs (hunk 5 above)
Comment (llogiq) [inline libimagentrytag/src/util.rs]: Yes.

File: imag-store/src/create.rs (hunk 2 above)
Comment (matthiasbeyer) [inline imag-store/src/create.rs]: Yes, would be nice if you'd remove it, please!

**Comment by llogiq** [comment]
Done (hopefully) :smile:

**Pull request merged**